akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -193,173 +269,63 @@ private[akka] class AffinityPool(
-
-  private final class ThreadPoolWorker(val q: BoundedTaskQueue, val idleStrategy: IdleStrategy) extends Runnable {
-
-    private sealed trait WorkerState
-    private case object NotStarted extends WorkerState
-    private case object InExecution extends WorkerState
-    private case object Idle extends WorkerState
-
-    val thread: Thread = tf.newThread(this)
-    @volatile private var workerState: WorkerState = NotStarted
-
-    def startWorker(): Unit = {
-      workerState = Idle
-      thread.start()
-    }
-
-    private def runCommand(command: Runnable) = {
-      workerState = InExecution
+    private[this] final def runCommand(command: Runnable): Unit = {
+      executing = true
       try
         command.run()
       finally
-        workerState = Idle
+        executing = false
     }
-
-    override def run(): Unit = {
-
+    override final def run(): Unit = {
       /**
-       * Determines whether the worker can keep running or not.
-       * In order to continue polling for tasks three conditions
-       * need to be satisfied:
-       *
-       * 1) pool state is less than Shutting down or queue
-       * is not empty (e.g pool state is ShuttingDown but there are still messages to process)
-       *
-       * 2) the thread backing up this worker has not been interrupted
-       *
-       * 3) We are not in ShutDown state (in which we should not be processing any enqueued tasks)
+       * We keep running as long as we are Running
+       * or we're ShuttingDown but we still have tasks to execute,
+       * and we're not interrupted.
        */
-      def shouldKeepRunning =
-        (poolState < ShuttingDown || !q.isEmpty) &&
-          !Thread.interrupted() &&
-          poolState != ShutDown
-
-      var abruptTermination = true
-      try {
-        while (shouldKeepRunning) {
+      @tailrec def runLoop(): Unit = {
+        val ps = poolState // Single volatile read
+        if (((ps == Running) || (ps == ShuttingDown && !q.isEmpty)) && !Thread.interrupted()) {
           val c = q.poll()
           if (c ne null) {
             runCommand(c)
             idleStrategy.reset()
-          } else // if not wait for a bit
-            idleStrategy.idle()
+          } else {
+            idleStrategy.idle() // if not wait for a bit
+          }
+          runLoop()
         }
+      }
+
+      var abruptTermination = true
+      try {
+        runLoop()
         abruptTermination = false // if we have reached here, our termination is not due to an exception
       } finally {
         onWorkerExit(this, abruptTermination)
       }
     }
 
-    def stop() =
-      if (!thread.isInterrupted && workerState != NotStarted)
-        thread.interrupt()
+    def stop(): Unit = if (!thread.isInterrupted) thread.interrupt()
 
-    def stopIfIdle() =
-      if (workerState == Idle)
-        stop()
+    def stopIfIdle(): Unit = if (!executing) stop()
```

Comment:
checking `executing` is a possible racy operation since it is check-then-act, I assume we're fine with that.